File: frontend/src/types/organization.ts
/**
 * Organization and multi-tenancy types for Knowledge Horizon
 *
 * Organized to mirror backend schemas/organization.py for easy cross-reference.
 * Section order:
 *   1. Re-exports (for backwards compatibility)
 *   2. Organization Types
 *   3. Member Types
 *   4. Subscription Types
 *   5. Notes Types
 *   6. Invitation Types
 */

// ============================================================================
// RE-EXPORTS (for backwards compatibility)
// ============================================================================

// Re-export user types that are commonly imported from organization.ts
export type { UserRole, OrgMember, User as AdminUser, UserList } from './user';

// ============================================================================
// ORGANIZATION TYPES
// ============================================================================

// Stream scope
export type StreamScope = 'global' | 'organization' | 'personal';

// Organization
export interface Organization {
  org_id: number;
  name: string;
  is_active: boolean;
  created_at: string;
  updated_at?: string;
}

// ============================================================================
// MEMBER TYPES
// ============================================================================

export interface OrganizationWithStats extends Organization {
  member_count: number;
  stream_count: number;
  pending_invitation_count: number;
}

export interface OrganizationUpdate {
  name?: string;
  is_active?: boolean;
}

// Member update
export interface OrgMemberUpdate {
  role: import('./user').UserRole;
}

// ============================================================================
// SUBSCRIPTION TYPES
// ============================================================================

export interface StreamSubscriptionStatus {
  stream_id: number;
  stream_name: string;
  scope: StreamScope;
  purpose?: string;
  is_org_subscribed?: boolean;
  is_user_subscribed: boolean;
  is_user_opted_out: boolean;
  created_at: string;
}

export interface GlobalStreamLibrary {
  streams: StreamSubscriptionStatus[];
  total_count: number;
}

export interface OrgStreamList {
  streams: StreamSubscriptionStatus[];
  total_count: number;
}

// ============================================================================
// NOTES TYPES
// ============================================================================

export interface ArticleNote {
  id: string;
  user_id: number;
  author_name: string;
  content: string;
  visibility: 'personal' | 'shared';
  created_at: string;
  updated_at: string;
}

export interface ArticleNoteCreate {
  content: string;
  visibility?: 'personal' | 'shared';
}

export interface ArticleNoteUpdate {
  content?: string;
  visibility?: 'personal' | 'shared';
}

export interface ArticleNotesResponse {
  report_id: number;
  article_id: number;
  notes: ArticleNote[];
  total_count: number;
}

// ============================================================================
// INVITATION TYPES
// ============================================================================

export interface Invitation {
  invitation_id: number;
  email: string;
  org_id?: number;
  org_name?: string;
  role: string;
  token: string;
  invite_url: string;
  created_at: string;
  expires_at: string;
  accepted_at?: string;
  is_revoked: boolean;
  inviter_email?: string;
}

File: backend/schemas/organization.py
"""
Organization schemas for multi-tenancy support.
"""

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime

from models import UserRole


# ============================================================================
# ORGANIZATION TYPES
# ============================================================================

class OrganizationBase(BaseModel):
    """Base organization fields."""
    name: str = Field(..., min_length=1, max_length=255, description="Organization name")


class OrganizationCreate(OrganizationBase):
    """Schema for creating an organization."""
    pass


class OrganizationUpdate(BaseModel):
    """Schema for updating an organization."""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    is_active: Optional[bool] = None


class Organization(OrganizationBase):
    """Full organization response."""
    org_id: int
    is_active: bool = True
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class OrganizationWithStats(Organization):
    """Organization with member counts."""
    member_count: int = 0
    pending_invitation_count: int = 0


# ============================================================================
# MEMBER TYPES
# ============================================================================

class OrgMember(BaseModel):
    """Organization member info."""
    user_id: int
    email: str
    full_name: Optional[str] = None
    role: UserRole
    joined_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class OrgMemberUpdate(BaseModel):
    """Schema for updating a member's role."""
    role: UserRole = Field(..., description="New role for the member")


class OrgMemberInvite(BaseModel):
    """Schema for inviting a new member."""
    email: str = Field(..., description="Email of the user to invite")
    role: UserRole = Field(default=UserRole.MEMBER, description="Role to assign")


# ============================================================================
# INVITATION TYPES
# ============================================================================

class Invitation(BaseModel):
    """Invitation to join an organization."""
    invitation_id: int
    email: str
    org_id: Optional[int] = None
    org_name: Optional[str] = None
    role: str
    token: str
    invite_url: str
    created_at: datetime
    expires_at: datetime
    accepted_at: Optional[datetime] = None
    is_revoked: bool = False
    inviter_email: Optional[str] = None

    class Config:
        from_attributes = True
